src/SigasiProjectCreator/ProjectCreator.py

```diff
@@ -0,0 +1,467 @@
+            self.check_and_create_virtual_folder(rel_path)
+            self.sigasi_project.add_link(rel_path, get_rel_or_abs_path(path, self.project_root, self.options), False)
+            abs_to_rel_file[path] = rel_path
+
+        if self.options.mapping == 'file':
+            self.create_library_mapping_per_file(entries, abs_to_rel_file)
+        else:
+            self.create_library_mapping_folders(entries, abs_to_rel_file)
+
+        return has_vhdl, has_verilog
+
+
+class ProjectCreatorLinkedFolders(ProjectCreator):
+    def __init__(self, options):
+        # super(ProjectCreatorLinkedFolders, self).__init__(options)
+        super().__init__(options)
+
+    def create_project_layout(self, entries):
+        has_vhdl = False
+        has_verilog = False
+
+        # design_folders is a list of folders with actual design files
+        design_folders = get_design_folders(entries)
+        design_root = get_design_root_folder(design_folders)
+        design_subtrees = get_design_subtrees(design_folders)
+        for subtree in design_subtrees:
+            self.check_and_create_linked_folder(pathlib.Path(os.path.relpath(subtree, design_root)), subtree)
+
+        abs_to_rel_file = {}
+        for path, library in entries.items():
+            assert not path.is_relative_to(self.project_root), \
+                f'*ERROR* linked project: destination folder {self.project_root} may not contain design file {path}'
+            file_ext = path.suffix.lower()
+            if file_ext in ['.vhd', '.vhdl']:
+                has_vhdl = True
+            elif file_ext in ['.v', '.sv']:
+                has_verilog = True
+            # abs_to_rel_file[path] = pathlib.Path(os.path.relpath(path))
+            for subtree in design_subtrees:
+                if path.is_relative_to(subtree):
+                    abs_to_rel_file[path] = path.relative_to(subtree.parent)
+
+        if self.options.mapping == 'file':
+            self.create_library_mapping_per_file(entries, abs_to_rel_file)
+        else:
+            self.create_library_mapping_folders(entries, abs_to_rel_file)
+
+        return has_vhdl, has_verilog
+
+
+def get_project_creator(options):
+    subclasses = {
+        'in-place': ProjectCreatorInPlace,
+        'simulator': ProjectCreatorSimulator,
+        'linked-files-flat': ProjectCreatorLinkedFilesFlat,
+        'linked-files-tree': ProjectCreatorLinkedFilesTree,
+        'linked-folders': ProjectCreatorLinkedFolders
+    }
+    assert options.layout in subclasses.keys(), f'Invalid layout option: {options.layout}'
+    return subclasses[options.layout](options)
+
+
+def get_parser_for_type(input_type):
+    if input_type == 'dotf':
+        return DotFfileParser.parse_file
+    if input_type == 'csv':
+        return CsvParser.parse_file
+    if input_type == 'hdp':
+        return parse_hdp_file
+    if input_type == 'filelist':
+        return None
+    if input_type == 'xise':
+        return parse_xilinx_file
```

Comment:
This should use a similar mechanism as the project creator, returning a class rather than a function (or have the current function return a dataclass or somewhat documented type, right now I have to go and read all of these functions to know what they could possibly return).

Reply:
Implemented